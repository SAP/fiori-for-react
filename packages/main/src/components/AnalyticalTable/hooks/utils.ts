export const makeTemplateColumns = (columns, resizedColumns) =>
  columns
    .map((column) => {
      if (resizedColumns[column.id]) {
        return `${resizedColumns[column.id]}px`;
      }
      return `minmax(${column.minWidth}px, ${column.width || '1fr'})`;
    })
    .join(' ');
