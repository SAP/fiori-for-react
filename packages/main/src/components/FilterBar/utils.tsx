import { cloneElement, ReactElement } from 'react';

export const setPropsOfChildren = (children, currentRef) => {
  return children.map((child) => {
    if (child.props.children.props[currentRef]?.children) {
      const childrenOfChild = child.props.children;
      if (
        child.props.children.props[currentRef]?.tagName === 'UI5-SELECT' ||
        child.props.children.props[currentRef]?.tagName === 'UI5-MULTI-COMBOBOX'
      ) {
        const selectedIndices = Array.from(child.props.children.props[currentRef].children)
          .map((item, index) => (item._state.selected ? index : false))
          .filter((el) => el !== false);
        const options = child.props.children.props.children.map((item, index) => {
          if (selectedIndices.includes(index)) {
            return cloneElement(item, { selected: true });
          }
          return cloneElement(item, { selected: false });
        });
        return cloneElement(child, {
          children: { ...childrenOfChild, props: { ...childrenOfChild.props, children: options } }
        });
      }
      if (
        child.props.children.props[currentRef]?.tagName === 'UI5-INPUT' ||
        child.props.children.props[currentRef]?.tagName === 'UI5-DATEPICKER'
      ) {
        return cloneElement(child, {
          children: {
            ...childrenOfChild,
            props: { ...childrenOfChild.props, value: child.props.children.props[currentRef]._state.value }
          }
        });
      }
      if (
        child.props.children.props[currentRef]?.tagName === 'UI5-SWITCH' ||
        child.props.children.props[currentRef]?.tagName === 'UI5-CHECKBOX'
      ) {
        return cloneElement(child, {
          children: {
            ...childrenOfChild,
            props: { ...childrenOfChild.props, checked: child.props.children.props[currentRef]._state.checked }
          }
        });
      }
      return child;
    }
    return child;
  });
};

export const addRef = (children, refsArray, currentRef) =>
  children.map((child) => {
    const childRef = refsArray.filter((item) => item.key === child.key);
    const childrenAttributes = child.props.children;
    const childrenProps = child.props.children.props;
    return cloneElement(child as ReactElement<any>, {
      children: { ...childrenAttributes, props: { ...childrenProps, [currentRef]: childRef[0]?.node } }
    });
  });

export const renderSearchWithValue = (searchElement, searchValue) => {
  return cloneElement(searchElement(), { value: searchValue });
};
