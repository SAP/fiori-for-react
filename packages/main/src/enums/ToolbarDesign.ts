export enum ToolbarDesign {
  PageFooter = 'PageFooter',
  ContainerBar = 'ContainerBar',
  ContentBar = 'ContentBar',
  ContentBarTransparent = 'ContentBarTransparent'
}
