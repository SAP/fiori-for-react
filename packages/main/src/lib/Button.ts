import { Button, ButtonPropTypes } from '../webComponents/Button';

export { Button, ButtonPropTypes };
