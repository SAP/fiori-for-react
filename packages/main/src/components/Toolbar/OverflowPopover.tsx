import React, { useEffect, useRef, useState } from 'react';
import { ToggleButton } from '@ui5/webcomponents-react/lib/ToggleButton';
import { ButtonDesign } from '@ui5/webcomponents-react/lib/ButtonDesign';
import { Popover } from '@ui5/webcomponents-react/lib/Popover';
import { PlacementType } from '@ui5/webcomponents-react/lib/PlacementType';
import '@ui5/webcomponents-icons/dist/icons/overflow';

export function OverflowPopover(props) {
  const { lastVisibleIndex, contentClass, children } = props;
  const popoverRef = useRef();
  const [pressed, setPressed] = useState(false);

  const handleToggleButtonClick = (e) => {
    if (popoverRef.current) {
      if (!pressed) {
        popoverRef.current.openBy(e.target);
        setPressed(true);
      } else {
        popoverRef.current.close();
      }
    }
  };

  const handleClose = () => {
    setPressed(false);
  };

  return (
    <>
      <ToggleButton
        design={ButtonDesign.Transparent}
        icon="sap-icon://overflow"
        onClick={handleToggleButtonClick}
        pressed={pressed}
      />
      <Popover placementType={PlacementType.Bottom} ref={popoverRef} onAfterClose={handleClose}>
        <div className={contentClass}>
          {React.Children.map(children, (item, index) => {
            if (index > lastVisibleIndex) return item;
            return null;
          })}
        </div>
      </Popover>
    </>
  );
}
