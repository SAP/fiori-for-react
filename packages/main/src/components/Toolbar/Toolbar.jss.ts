import { ThemingParameters } from '@ui5/webcomponents-react-base/lib/ThemingParameters';
import { CssSizeVariables } from '@ui5/webcomponents-react-base/lib/CssSizeVariables';
export const styles = {
  outerContainer: {
    width: '100%',
    //todo
    // minWidth: '50px',
    height: CssSizeVariables.sapWcrToolbarHeight,
    position: 'relative',
    overflow: 'hidden',
    display: 'flex',
    justifyContent: 'flex-end',
    alignItems: 'center',
    borderBottom: `solid 0.0625rem ${ThemingParameters.sapGroup_TitleBorderColor}`
  },
  //todo do this in index.tsx
  active: {
    '&:active': {
      background: `${ThemingParameters.sapActiveColor}`
    }
  },
  activeInfo: {
    '&:active': {
      background: `${ThemingParameters.sapInfobar_Active_Background}`
    }
  },
  clear: {
    borderBottom: 'none'
  },
  info: {
    //todo not active background color: darken(@sapUiBaseColor, 10);
    background: ThemingParameters.sapInfobar_Background,
    '&:hover': {
      background: ThemingParameters.sapInfobar_Hover_Background
    }
  },
  solid: {
    background: ThemingParameters.sapBackgroundColor,
    '&:hover': {
      //todo hover color contrast(@sapUiBaseBG, darken(@sapUiBaseBG, 3), lighten(@sapUiBaseBG, 3));
      background: 'lightblue'
    }
  },
  transparent: {
    background: ThemingParameters.sapToolbar_Background,
    '&:hover': {
      //todo hover color contrast(@sapUiToolbarBackground, darken(@sapUiToolbarBackground, 3), lighten(@sapUiToolbarBackground, 3));
      background: 'lightgrey'
    },
    '&:active': {
      background: ThemingParameters.sapActiveColor
    }
  },
  toolbar: {
    width: '100%',
    '& >:first-child:not(.spacer)': {
      margin: '0 0.25rem 0 0'
    },
    '& >:last-child:not(.spacer)': {
      margin: '0 0.5rem 0 0.25rem'
    },
    '& > *:not(first-child):not(last-child):not(.spacer)': {
      margin: '0 0.25rem'
    },
    display: 'flex',
    alignItems: 'center',
    position: 'absolute',
    top: 0,
    left: 0,
    bottom: 0
  },
  overflowButtonContainer: {
    marginRight: '0.5rem'
  },
  popoverContent: {
    maxWidth: '20rem',
    padding: CssSizeVariables.sapWcrToolbarPopoverContentPadding,
    display: 'flex',
    flexDirection: 'column'
  }
};
