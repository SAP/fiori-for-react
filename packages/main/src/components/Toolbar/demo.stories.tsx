import React from 'react';
import { Toolbar } from './index';
import { Button } from '../../webComponents/Button';
import { text } from '@storybook/addon-knobs';

export const renderStory = () => {
  return (
    <Toolbar width={text('width', '50%')}>
      <Button>Item1</Button>
      <Button>Item2</Button>
      <Button>Item3</Button>
      <Button>Item4</Button>
      <Button>Item5</Button>
      <Button>Item6</Button>
      <Button>Item7</Button>
      <Button>Item8</Button>
      <Button>Item9</Button>
      <Button>Item10</Button>
      <Button>Item11</Button>
      <Button>Item12</Button>
    </Toolbar>
  );
};

renderStory.story = {
  name: 'Default'
};

export default {
  title: 'Components | Toolbar',
  component: Toolbar
};
