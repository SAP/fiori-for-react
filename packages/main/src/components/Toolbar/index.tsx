import { createComponentStyles } from '@ui5/webcomponents-react-base/lib/createComponentStyles';
import { StyleClassHelper } from '@ui5/webcomponents-react-base/lib/StyleClassHelper';
import { useConsolidatedRef } from '@ui5/webcomponents-react-base/lib/useConsolidatedRef';
import { enrichEventWithDetails } from '@ui5/webcomponents-react-base/lib/Utils';
import { CommonProps } from '@ui5/webcomponents-react/interfaces/CommonProps';
import { ToolbarDesign } from '@ui5/webcomponents-react/lib/ToolbarDesign';
import { ToolbarStyle } from '@ui5/webcomponents-react/lib/ToolbarStyle';
import { usePassThroughHtmlProps } from '@ui5/webcomponents-react-base/lib/usePassThroughHtmlProps';
import React, {
  createRef,
  FC,
  forwardRef,
  ReactNode,
  ReactNodeArray,
  Ref,
  RefObject,
  useCallback,
  useEffect,
  useLayoutEffect,
  useMemo,
  useRef,
  useState
} from 'react';
import ResizeObserver from 'resize-observer-polyfill';
import { OverflowPopover } from './OverflowPopover';
import { styles } from './Toolbar.jss';

const useStyles = createComponentStyles(styles, { name: 'Toolbar' });

export interface ToolbarProptypes extends CommonProps {
  children?: ReactNode | ReactNodeArray;
  toolbarStyle?: ToolbarStyle;
  design?: ToolbarDesign;
  active?: boolean;
  onToolbarClick?: (event: CustomEvent<{}>) => void;
}

const Toolbar: FC<ToolbarProptypes> = forwardRef((props: ToolbarProptypes, ref: Ref<HTMLDivElement>) => {
  const { children, toolbarStyle, design, active, style, tooltip, className, onToolbarClick } = props;
  const classes = useStyles(styles);
  const outerContainer: RefObject<HTMLDivElement> = useConsolidatedRef(ref);
  const controlMetaData = useRef([]);
  const [lastVisibleIndex, setLastVisibleIndex] = useState(null);
  const [blockerWidth, setBlockerWidth] = useState(0);
  const toolbarClasses = StyleClassHelper.of(classes.outerContainer);

  const passThroughProps = usePassThroughHtmlProps(props, ['onToolbarClick']);

  if (toolbarStyle === ToolbarStyle.Clear) {
    toolbarClasses.put(classes.clear);
  }

  if (active) {
    toolbarClasses.put(classes.active);
  }

  switch (design) {
    case ToolbarDesign.Solid:
      toolbarClasses.put(classes.solid);
      break;
    case ToolbarDesign.Transparent:
      toolbarClasses.put(classes.transparent);
      break;
    case ToolbarDesign.Info:
      if (active) {
        toolbarClasses.put(classes.activeInfo);
      } else {
        toolbarClasses.put(classes.info);
      }
      break;
    default:
      break;
  }

  if (className) {
    toolbarClasses.put(className);
  }

  const childrenWithRef = useMemo(() => {
    controlMetaData.current = [];

    return React.Children.toArray(children).map((item, index) => {
      const itemRef: RefObject<HTMLDivElement> = createRef();

      controlMetaData.current.push({
        ref: itemRef
      });
      if (item.type.displayName === 'ToolbarSpacer') {
        return item;
      }
      return (
        <div ref={itemRef} key={index}>
          {item}
        </div>
      );
    });
  }, [children]);
  const overflowNeeded =
    (lastVisibleIndex || lastVisibleIndex === 0) && React.Children.count(childrenWithRef) !== lastVisibleIndex + 1;

  const calculateVisibleItems = useCallback(() => {
    requestAnimationFrame(() => {
      if (!outerContainer.current) return;
      const availableWidth = outerContainer.current.getBoundingClientRect().width; /*- 32 - 8*/
      let consumedWidth = 0;
      let lastIndex = null;
      let lastFitWidth = 0;

      if (availableWidth - 32 - 8 <= 0) {
        lastIndex = -1;
        lastFitWidth = 0;
      } else {
        controlMetaData.current.forEach((item, index) => {
          const currentMeta = controlMetaData.current[index];
          if (currentMeta && currentMeta.ref && currentMeta.ref.current) {
            let nextWidth = currentMeta.ref.current.getBoundingClientRect().width;
            nextWidth += index === 0 || index === controlMetaData.current.length - 1 ? 4 : 8; //first & last element = padding: 4px
            if (index === controlMetaData.current.length - 1) {
              if (consumedWidth + nextWidth <= availableWidth - 8) {
                lastIndex = index;
                lastFitWidth = consumedWidth + nextWidth;
              }
            } else {
              if (consumedWidth + nextWidth <= availableWidth - 32 - 8) {
                lastIndex = index;
                lastFitWidth = consumedWidth + nextWidth;
              }
              if (consumedWidth < availableWidth - 32 - 8 && consumedWidth + nextWidth >= availableWidth - 32 - 8) {
                lastIndex = index - 1;
                lastFitWidth = 0;
              }
            }
            consumedWidth += nextWidth;
          }
        });
      }
      setBlockerWidth(Math.max(0, availableWidth - lastFitWidth));
      setLastVisibleIndex(lastIndex);
    });
  }, [outerContainer.current, controlMetaData.current, setLastVisibleIndex, childrenWithRef, overflowNeeded]);

  const renderBlocker = useCallback(
    () => <div data-toolbar-blocker style={{ width: `${blockerWidth}px`, minWidth: `${blockerWidth}px` }} />,
    [blockerWidth]
  );

  const observer = useRef(new ResizeObserver(calculateVisibleItems));

  useEffect(() => {
    if (outerContainer.current) {
      observer.current.observe(outerContainer.current);
    }
    return () => {
      observer.current.disconnect();
    };
  }, [outerContainer.current]);

  useLayoutEffect(() => {
    calculateVisibleItems();
  }, [calculateVisibleItems]);

  const renderOverflowPopover = useCallback(() => {
    return (
      <div className={classes.overflowButtonContainer} title="Show More">
        <OverflowPopover lastVisibleIndex={lastVisibleIndex} contentClass={classes.popoverContent}>
          {children}
        </OverflowPopover>
      </div>
    );
  }, [classes.popoverContent, lastVisibleIndex, children]);

  const onClick = useCallback(
    (e) => {
      if (active && onToolbarClick) {
        if (onToolbarClick) {
          onToolbarClick(enrichEventWithDetails(e));
        }
      }
    },
    [onToolbarClick, active]
  );

  return (
    <div
      title={tooltip}
      style={style}
      className={toolbarClasses.toString()}
      ref={outerContainer}
      onClick={onClick}
      {...passThroughProps}
    >
      <div className={classes.toolbar}>
        {overflowNeeded &&
          React.Children.map(childrenWithRef, (item, index) => {
            if (index === lastVisibleIndex) {
              return (
                <>
                  {item}
                  {renderBlocker()}
                </>
              );
            }
            if (lastVisibleIndex === -1) {
              return (
                <>
                  {renderBlocker()}
                  {item}
                </>
              );
            }
            return item;
          })}
        {!overflowNeeded && childrenWithRef}
      </div>
      {overflowNeeded && renderOverflowPopover()}
    </div>
  );
});

Toolbar.defaultProps = {
  toolbarStyle: ToolbarStyle.Standard,
  design: ToolbarDesign.Auto,
  active: false
};

Toolbar.displayName = 'Toolbar';
export { Toolbar };
