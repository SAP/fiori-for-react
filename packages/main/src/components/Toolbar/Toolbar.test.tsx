import { createPassThroughPropsTest } from '@shared/tests/utils';
import { cleanup, render, screen, waitFor, fireEvent } from '@testing-library/react';
import { Text } from '@ui5/webcomponents-react/lib/Text';
import { Toolbar } from '@ui5/webcomponents-react/lib/Toolbar';
import { ToolbarDesign } from '@ui5/webcomponents-react/lib/ToolbarDesign';
import { ToolbarStyle } from '@ui5/webcomponents-react/lib/ToolbarStyle';
import React from 'react';
import { act } from 'react-dom/test-utils';

describe('Toolbar', () => {
  test('Renders with default Props', () => {
    expect(render(<Toolbar />).asFragment()).toMatchSnapshot();
  });

  test('Renders with children', () => {
    expect(
      render(
        <Toolbar>
          <Text>Item1</Text>
          <Text>Item2</Text>
          <Text>Item3</Text>
        </Toolbar>
      ).asFragment()
    ).toMatchSnapshot();
    expect(screen.getByText('Item1').textContent).toEqual('Item1');
  });

  // test('Renders overflowButton', async () => {
  //   let utils;
  //
  //   await act(async () => {
  //     utils = render(
  //       <Toolbar style={{ width: '50px' }}>
  //         <Text>Item1</Text>
  //         <Text>Item2</Text>
  //         <Text>Item3</Text>
  //       </Toolbar>
  //     );
  //   });
  //
  //   expect(utils.asFragment()).toMatchSnapshot();
  //
  //   await waitFor(() => expect(screen.getByTitle('Show More').className).toMatch(/overflowButtonContainer/));
  // });

  test('Design Style modes', () => {
    const { rerender } = render(
      <Toolbar tooltip={'toolbar'} toolbarStyle={ToolbarStyle.Clear}>
        <Text>Item1</Text>
        <Text>Item2</Text>
        <Text>Item3</Text>
      </Toolbar>
    );
    expect(screen.getByTitle('toolbar').className).toMatch(/clear/);

    rerender(
      <Toolbar tooltip={'toolbar'} toolbarStyle={ToolbarStyle.Clear} design={ToolbarDesign.Transparent}>
        <Text>Item1</Text>
        <Text>Item2</Text>
        <Text>Item3</Text>
      </Toolbar>
    );
    expect(screen.getByTitle('toolbar').className).toMatch(/transparent/);

    rerender(
      <Toolbar tooltip={'toolbar'} toolbarStyle={ToolbarStyle.Clear} design={ToolbarDesign.Solid}>
        <Text>Item1</Text>
        <Text>Item2</Text>
        <Text>Item3</Text>
      </Toolbar>
    );
    expect(screen.getByTitle('toolbar').className).toMatch(/solid/);

    rerender(
      <Toolbar tooltip={'toolbar'} toolbarStyle={ToolbarStyle.Clear} design={ToolbarDesign.Info}>
        <Text>Item1</Text>
        <Text>Item2</Text>
        <Text>Item3</Text>
      </Toolbar>
    );
    expect(screen.getByTitle('toolbar').className).toMatch(/info/);

    rerender(
      <Toolbar tooltip={'toolbar'} toolbarStyle={ToolbarStyle.Clear} active design={ToolbarDesign.Info}>
        <Text>Item1</Text>
        <Text>Item2</Text>
        <Text>Item3</Text>
      </Toolbar>
    );
    expect(screen.getByTitle('toolbar').className).toMatch(/activeInfo/);

    rerender(
      <Toolbar tooltip={'toolbar'} toolbarStyle={ToolbarStyle.Clear} active>
        <Text>Item1</Text>
        <Text>Item2</Text>
        <Text>Item3</Text>
      </Toolbar>
    );
    expect(screen.getByTitle('toolbar').className).toMatch(/active/);
  });

  createPassThroughPropsTest(Toolbar);
});
