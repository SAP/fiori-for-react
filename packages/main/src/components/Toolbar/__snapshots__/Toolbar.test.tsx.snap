// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toolbar Renders with children 1`] = `
<DocumentFragment>
  <div
    class="Toolbar-outerContainer-0-2-1"
  >
    <div
      class="Toolbar-toolbar-0-2-9"
    >
      <div>
        <span
          class="Text-text-0-2-12"
        >
          Item1
        </span>
      </div>
      <div>
        <span
          class="Text-text-0-2-12"
        >
          Item2
        </span>
      </div>
      <div>
        <span
          class="Text-text-0-2-12"
        >
          Item3
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Toolbar Renders with default Props 1`] = `
<DocumentFragment>
  <div
    class="Toolbar-outerContainer-0-2-1"
  >
    <div
      class="Toolbar-toolbar-0-2-9"
    />
  </div>
</DocumentFragment>
`;
